### PR TITLE
Architecture overview

### PR DESCRIPTION
**Pull request by TheOnlyError**
Adds the architecture document as proposed by #2524.

This PR serves as a place for discussion regarding updates, suggestions and improvements :smiley:.

### CODE REVIEW

File: architecture/architecture.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+```
+
+### Compiler
+
+Python compilation to bytecode. The interface is exposed through the porcelain crate with `compile` and `compile_symtable`, while the inner workings are defined in compiler/src/compile.rs, which is mostly an adaptation of the CPython implementation.
+
+### Derive
+
+Rust language extensions and macros specific to rustpython. Here we can find the definition of `PyModule` and `PyClass` along with usefule macros like `py_compile!`
```
Comment (youknowone) [inline architecture/architecture.md:63]:
```suggestion
Rust language extensions and macros specific to rustpython. Here we can find the definition of `PyModule` and `PyClass` along with useful macros like `py_compile!`
```

File: architecture/architecture.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+- The interface and helper for the REPL are defined in this package, but the actual REPL can be found in `vm/rustyline.rs`
+
+
+### WASM
+
+Crate for WebAssembly build, which compiles the RustPython package to a format that can be run on any modern browser.
+
+### extra_tests
+
+Integration ans snippet tests for the entire project.
```
Comment (KipHamiltons) [inline architecture/architecture.md:115]:
```suggestion
Integration and snippet tests for the entire project.
```

**Comment by DimitrisJim** [comment]
Wouldn't this be a suitable document to be placed on the rustpython web page? The docs repo looks like a good candidate.

**Comment by youknowone** [comment]
@DimitrisJim I agree. Would you please review it?

File: architecture/architecture.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+- A test has been marked as `unittest.expectedFailure` with `TODO: RustPython` left on top if the test can run but failed
+
+Note: This is a recommended route to starting with contributing. To get started please take a look [this blog post](https://rustpython.github.io/guideline/2020/04/04/how-to-contribute-by-cpython-unittest.html).
+
+
+### VM
+
+Python VM
+
+- builtins: all the builtin functions
```
Comment (DimitrisJim) [inline architecture/architecture.md:96]: Builtin functions live in `make_module.rs` of `vm/src/builtins`. Rest of the files in `builtins` contain the builtin types. Not sure what `obj` is, an older folder that used to exist?

File: architecture/architecture.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+- obj: Builtin types
+- stdlib: the parts of the standard library implemented in rust
+
+### src
+
+The RustPython executable is implemented here, which is the interface through which users come in contact with library.
+Some things to note:
+
+- The CLI is defined in `src/lib.rs@run`
+- The interface and helper for the REPL are defined in this package, but the actual REPL can be found in `vm/rustyline.rs`
```
Comment (DimitrisJim) [inline architecture/architecture.md:106]: it's `vm/src/readline.rs` (now, at least).

File: architecture/architecture.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+### Lib
+
+Python side of the standard libary, copied over (with care) from CPython sourcecode.
+
+### Lib/test
+
+
+CPython test suite, which can be used to compare with CPython in terms of functionality and performance.
+Many of these files have been modified to fit with the current state of RustPython (when they were added), with one of three ways:
```
Comment (DimitrisJim) [inline architecture/architecture.md:83]:
```suggestion
Many of these files have been modified to fit with the current state of RustPython (when they were added), in one of three ways:
```

File: architecture/architecture.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+This module (single file at moment) holds the representation of bytecode for RustPython
+<!-- For instance, the following function
+```python
+def f(x):
+    return x + 1
+```
+Is compiled to
+```rust
+// Not sure what this should be yet
```
Comment (DimitrisJim) [inline architecture/architecture.md:56]: As in, what the bytecode should be? You can currently `dis` the function:

```
>>>>> dis(f)
                 0 LoadFast             (0, x)
                 1 LoadConst            (1)
                 2 BinaryOperation      (Add)
                 3 ReturnValue
```

**Review by DimitrisJim — APPROVED** [review]
finally got to this :smile: I think this does offer value in this repo, thanks (very belatedly!) for the PR!

File: architecture/architecture.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
+# Architecture
+
+This document contains a high-level architectural overview of RustPython, thus it's very well-suited to get to know [the codebase][1].
+
+RustPython is an Open Source (MIT) Python 3 interpreter written in Rust, available as both a library and a shell environment. Using Rust to implement the Python interpreter enables Python to be used as a programming language for Rust applications. Moreover, it allows Python to be immediately compiled in the browser using WebAssembly, meaning that anyone could easily run their Python code in the browser. For a more detailed introduction to RustPython, have a look at [this blog post][2].
```
Comment (fanninpm) [inline architecture/architecture.md:5]:
```suggestion
RustPython is an Open Source (MIT-licensed) Python 3 interpreter written in Rust, available as both a library and a shell environment. Using Rust to implement the Python interpreter enables Python to be used as a programming language for Rust applications. Moreover, it allows Python to be immediately compiled in the browser using WebAssembly, meaning that anyone could easily run their Python code in the browser. For a more detailed introduction to RustPython, have a look at [this blog post][2].
```

Actually, I think the repo is dual-licensed under both the MIT/X11 license and the Apache license.

File: architecture/architecture.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+RustPython consists of several components which are described in the section below. Take a look at [this video][3] for a brief walk-through of the components of RustPython. For a more elaborate introduction to one of these components, the parser, see [this blog post][4] for more information.
+
+Have a look at these websites for a demo of RustPython running in the browser using WebAssembly:
+
+- [The demo page.][5]
+- [The RustPython Notebook, a toy notebook inspired by the Iodide project.][6]
+
+If, after reading this, you are to contribute to RustPython, take a look at these sources to get to know how and where to start:
```
Comment (fanninpm) [inline architecture/architecture.md:14]:
```suggestion
If, after reading this, you want to contribute to RustPython, take a look at these sources to get to know how and where to start:
```

This feels more friendly and approachable.

File: architecture/architecture.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+- [RustPython Development Guide and Tips][7]
+- [How to contribute to RustPython using CPython's unit tests][8]
+
+## Bird's eye view
+
+A high-level overview of the workings of RustPython is visible in the figure below, showing how Python source files are interpreted.
+
+![overview.png](overview.png)
```
Comment (fanninpm) [inline architecture/architecture.md:23]: You might want to consider [rewriting this diagram in Mermaid](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams).

File: architecture/architecture.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+Is compiled to the following bytecode:
+
+```python
+  2           0 LoadFast             (0, x)
+              1 LoadConst            (1)
+              2 BinaryOperation      (Add)
+              3 ReturnValue
+```
+
+Note that bytecode is subject to change, and is not a stable interface.
```
Comment (fanninpm) [inline architecture/architecture.md:65]:
```suggestion
Note that bytecode is subject to change, and is _not_ a stable interface.
```

I think this point is worth emphasizing.

**Review by fanninpm — COMMENTED** [review]
A few nitpicks, mostly non-blocking.

File: architecture/architecture.md (hunk 9 above)
Comment (DimitrisJim) [reply architecture/architecture.md:23]: that's wild. +1 but can probably leave this for a later point.

**Pull request merged**